### PR TITLE
Replace hclog.Logger with  LeveledLogger interface

### LINKED ISSUE #74: Introduce more generic Logger interface
Hey,
I am using [uber-go/zap](https://github.com/uber-go/zap) for logger and I created a wrapper that impements the basic Logger interface
```go
// Logger interface allows to use other loggers than
// standard log.Logger.
type Logger interface {
	Printf(string, ...interface{})
}
```
but I don't want to log debug logs, and in general I'd like more control.

I'd like to propose to change the `Logger` interface to something more generic so then every logger library can implement it
```go
type Logger interface {
	Errorf(format string, v ...interface{})
        Infof(format string, v ...interface{})
        Debugf(format string, v ...interface{})
        Warnf(format string, v ...interface{})
}
```
This can be introduced either with backwards compatible way by keeping the old Logger interface  or by breaking it as we are still on v0.X releases.

Another Option is to introduce an `ExtendedLogger` that is a generic interface that will replace the `hclog.Logger` interface in backwards compatible way.
```go
// ExtendedLogger interface implements the basic methods that a logger library needs
type ExtendedLogger interface {
	Error(string, ...interface{})
	Info(string, ...interface{})
	Debug(string, ...interface{})
	Warn(string, ...interface{})
}
```

Please let me know what you think, more than happy to make the changes :) 

Cheers

### PR DESCRIPTION
**Pull request by geototti21**
## what was done
* Changing the  `hclog.Logger` interface to a smaller one that will be easier to get implemented from other logger libraries. As we are just reducing the required methods on the interface this will not break the API.

## alternative:
As proposed on #74 we can also break the API by changing the existing Logger interface.

Closes #74

### CODE REVIEW

**Comment by geototti21** [comment]
Hey @ryanuber @evanphx, could I get some feeback on this please 🙏

File: client.go [hunk 1]
```diff
@@ -483,7 +489,7 @@ func (c *Client) Do(req *Request) (*http.Response, error) {
 			case Logger:
 				c.RequestLogHook(v, req.Request, i)
 			case hclog.Logger:
```
Comment (ryanuber) [inline client.go]: This should also be updated to the `ExtendedLogger` if the intent is to allow the use of non-stdlib and non-hclog loggers to be used.

File: client.go [hunk 2]
```diff
@@ -240,14 +240,22 @@ type Logger interface {
...
+	Info(string, ...interface{})
+	Debug(string, ...interface{})
+	Warn(string, ...interface{})
+}
+
+// HookLogger adapts an ExtendedLogger to Logger for use by the existing hook functions
 // without changing the API.
-type hookLogger struct {
-	logger hclog.Logger
+type HookLogger struct {
```
[the remark this answers is not in the document]
Comment (geototti21) [reply client.go]: you are right, there is no need; I kept it unexported as before

File: client.go [hunk 3]
```diff
@@ -335,9 +343,7 @@ func (c *Client) logger() interface{} {
 		}
 
 		switch c.Logger.(type) {
-		case Logger:
-			// ok
-		case hclog.Logger:
+		case Logger, ExtendedLogger:
```
[the remark this answers is not in the document]
Comment (geototti21) [reply client.go]: updated

File: client.go (hunk 1 above)
Comment (geototti21) [reply client.go]: oops yeah, updated

**Comment by geototti21** [comment]
> Hey @geototti21 , this seems reasonable. I left a few comments I think we should address prior to merge.
> 
> I'm also a little on the fence about the name `ExtendedLogger`, because it's not actually an extension of `Logger` but rather a totally different interface. Perhaps something like `LeveledLogger` would be more appropriate, as the interface simply has a method per level rather than a single firehose `Printf` method.

naming is hard :see_no_evil: , I renamed it to `LeveledLogger` :+1:

**Comment by geototti21** [comment]
I adressed all the comments @ryanuber, let me know if it is okay now please :)

**Comment by chrboe** [comment]
_ping_, any news on this @ryanuber?

**Comment by ryanuber** [comment]
Sorry about the delay, I was away for a while. Looks good!

**Pull request merged**